File: backend/trafego_ai/tools/meta_ads_api.py
"""
Implementação da integração com a API do Meta ADS (Facebook)
"""
import time
import json
import logging
from datetime import datetime, timedelta
from facebook_business.api import FacebookAdsApi
from facebook_business.adobjects.adaccount import AdAccount
from facebook_business.adobjects.campaign import Campaign
from facebook_business.adobjects.adset import AdSet
from facebook_business.adobjects.ad import Ad
from facebook_business.adobjects.targetingsearch import TargetingSearch
from facebook_business.adobjects.adimage import AdImage
from facebook_business.adobjects.advideo import AdVideo
from facebook_business.exceptions import FacebookRequestError

from backend.trafego_ai.config.settings import (
    META_APP_ID,
    META_APP_SECRET,
    META_ACCESS_TOKEN,
    META_ACCOUNT_ID
)

# Configurar logger
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


class MetaAdsAPI:
    """
    Classe para interação com a API do Meta ADS (Facebook).
    Implementa métodos para criar e gerenciar campanhas, conjuntos de anúncios e anúncios.
    """

    def __init__(self, app_id=None, app_secret=None, access_token=None, account_id=None):
        """
        Inicializa a integração com a API do Meta ADS.
        
        Args:
            app_id (str, optional): ID da aplicação Meta. Default para o valor nas configurações.
            app_secret (str, optional): Segredo da aplicação Meta. Default para o valor nas configurações.
            access_token (str, optional): Token de acesso. Default para o valor nas configurações.
            account_id (str, optional): ID da conta publicitária. Default para o valor nas configurações.
        """
        self.app_id = app_id or META_APP_ID
        self.app_secret = app_secret or META_APP_SECRET
        self.access_token = access_token or META_ACCESS_TOKEN
        self.account_id = account_id or META_ACCOUNT_ID
        
        # Inicializar a API do Facebook
        try:
            FacebookAdsApi.init(self.app_id, self.app_secret, self.access_token)
            logger.info("Facebook Ads API inicializada com sucesso")
            
            # Validar a conta
            self.ad_account = AdAccount(f'act_{self.account_id}')
            self.check_account_access()
        except FacebookRequestError as e:
            logger.error(f"Erro ao inicializar a API do Facebook: {e}")
            raise
        except Exception as e:
            logger.error(f"Erro desconhecido ao inicializar a API do Facebook: {e}")
            raise
    
    def check_account_access(self):
        """
        Verifica se temos acesso à conta de anúncios.
        
        Returns:
            bool: True se o acesso for válido, False caso contrário
        """
        try:
            # Tenta acessar informações da conta para verificar as permissões
            self.ad_account.api_get(fields=['name', 'account_status'])
            return True
        except FacebookRequestError as e:
            logger.error(f"Erro ao acessar a conta: {e}")
            return False
    
    def criar_campanha(self, nome, objetivo, orcamento_diario=None, orcamento_lifetime=None, 
                      data_inicio=None, data_fim=None, status="PAUSED"):
        """
        Cria uma nova campanha publicitária no Meta ADS.
        
        Args:
            nome (str): Nome da campanha
            objetivo (str): Objetivo da campanha (Ex: REACH, TRAFFIC, CONVERSIONS)
            orcamento_diario (float, optional): Orçamento diário em centavos
            orcamento_lifetime (float, optional): Orçamento total da campanha em centavos
            data_inicio (datetime, optional): Data de início da campanha
            data_fim (datetime, optional): Data de término da campanha
            status (str, optional): Status inicial da campanha. Default para "PAUSED"
            
        Returns:
            str: ID da campanha criada
        """
        try:
            params = {
                'name': nome,
                'objective': objetivo,
                'status': status,
                'special_ad_categories': [],
            }
            
            # Configurar orçamento
            if orcamento_diario:
                params['daily_budget'] = int(orcamento_diario * 100)  # Converter para centavos
            elif orcamento_lifetime:
                params['lifetime_budget'] = int(orcamento_lifetime * 100)  # Converter para centavos
            
            # Configurar datas
            if data_inicio:
                params['start_time'] = data_inicio.strftime("%Y-%m-%dT%H:%M:%S%z")
            if data_fim:
                params['end_time'] = data_fim.strftime("%Y-%m-%dT%H:%M:%S%z")
            
            # Criar a campanha
            campaign = self.ad_account.create_campaign(params=params)
            campaign_id = campaign['id']
            logger.info(f"Campanha criada com sucesso: {campaign_id}")
            
            return campaign_id
        
        except FacebookRequestError as e:
            logger.error(f"Erro ao criar campanha: {e}")
            raise
    
    def criar_conjunto_anuncios(self, campanha_id, nome, objetivo_otimizacao, segmentacao, 
                              orcamento_diario=None, orcamento_lifetime=None, 
                              data_inicio=None, data_fim=None, status="PAUSED"):
        """
        Cria um novo conjunto de anúncios (Ad Set) no Meta ADS.
        
        Args:
            campanha_id (str): ID da campanha
            nome (str): Nome do conjunto de anúncios
            objetivo_otimizacao (str): Objetivo de otimização (Ex: REACH, LINK_CLICKS)
            segmentacao (dict): Configurações de segmentação
            orcamento_diario (float, optional): Orçamento diário em centavos
            orcamento_lifetime (float, optional): Orçamento total do conjunto em centavos
            data_inicio (datetime, optional): Data de início
            data_fim (datetime, optional): Data de término
            status (str, optional): Status inicial. Default para "PAUSED"
            
        Returns:
            str: ID do conjunto de anúncios criado
        """
        try:
            params = {
                'name': nome,
                'campaign_id': campanha_id,
                'optimization_goal': objetivo_otimizacao,
                'billing_event': 'IMPRESSIONS',  # ou LINK_CLICKS, APP_INSTALLS, etc.
                'status': status,
                'targeting': segmentacao,
            }
            
            # Configurar orçamento
            if orcamento_diario:
                params['daily_budget'] = int(orcamento_diario * 100)  # Converter para centavos
            elif orcamento_lifetime:
                params['lifetime_budget'] = int(orcamento_lifetime * 100)  # Converter para centavos
            
            # Configurar datas
            if data_inicio:
                params['start_time'] = data_inicio.strftime("%Y-%m-%dT%H:%M:%S%z")
            if data_fim:
                params['end_time'] = data_fim.strftime("%Y-%m-%dT%H:%M:%S%z")
            
            # Criar o conjunto de anúncios
            ad_set = self.ad_account.create_ad_set(params=params)
            ad_set_id = ad_set['id']
            logger.info(f"Conjunto de anúncios criado com sucesso: {ad_set_id}")
            
            return ad_set_id
        
        except FacebookRequestError as e:
            logger.error(f"Erro ao criar conjunto de anúncios: {e}")
            raise
    
    def criar_anuncio(self, conjunto_anuncios_id, nome, creative_id, status="PAUSED"):
        """
        Cria um novo anúncio no Meta ADS.
        
        Args:
            conjunto_anuncios_id (str): ID do conjunto de anúncios
            nome (str): Nome do anúncio
            creative_id (str): ID do criativo a ser usado
            status (str, optional): Status inicial. Default para "PAUSED"
            
        Returns:
            str: ID do anúncio criado
        """
        try:
            params = {
                'name': nome,
                'adset_id': conjunto_anuncios_id,
                'creative': {'creative_id': creative_id},
                'status': status,
            }
            
            # Criar o anúncio
            ad = self.ad_account.create_ad(params=params)
            ad_id = ad['id']
            logger.info(f"Anúncio criado com sucesso: {ad_id}")
            
            return ad_id
        
        except FacebookRequestError as e:
            logger.error(f"Erro ao criar anúncio: {e}")
            raise
    
    def criar_criativo(self, titulo, texto, cta, url_destino, imagem_url=None, imagem_hash=None, 
                      formato="LINK"):
        """
        Cria um novo criativo para anúncios no Meta ADS.
        
        Args:
            titulo (str): Título do anúncio
            texto (str): Texto principal do anúncio
            cta (str): Call-to-action (Ex: LEARN_MORE, SHOP_NOW)
            url_destino (str): URL de destino do anúncio
            imagem_url (str, optional): URL da imagem a ser usada
            imagem_hash (str, optional): Hash de uma imagem já carregada
            formato (str, optional): Formato do criativo. Default para "LINK"
            
        Returns:
            str: ID do criativo criado
        """
        try:
            # Verificar se temos uma imagem para usar
            if imagem_url and not imagem_hash:
                # Fazer upload da imagem
                image = AdImage(parent_id=f'act_{self.account_id}')
                image[AdImage.Field.filename] = imagem_url
                image.remote_create()
                imagem_hash = image[AdImage.Field.hash]
            
            params = {
                'name': f'Criativo - {titulo[:20]}',
                'object_story_spec': {
                    'page_id': '123456789',  # ID da Página do Facebook - deve ser substituído pelo ID real
                    'link_data': {
                        'message': texto,
                        'link': url_destino,
                        'name': titulo,
                        'call_to_action': {'type': cta},
                    }
                }
            }
            
            # Adicionar a imagem ao criativo
            if imagem_hash:
                params['object_story_spec']['link_data']['image_hash'] = imagem_hash
            
            # Criar o criativo
            creative = self.ad_account.create_ad_creative(params=params)
            creative_id = creative['id']
            logger.info(f"Criativo criado com sucesso: {creative_id}")
            
            return creative_id
        
        except FacebookRequestError as e:
            logger.error(f"Erro ao criar criativo: {e}")
            raise
    
    def obter_metricas_campanha(self, campanha_id, data_inicio=None, data_fim=None, 
                               metricas=None):
        """
        Obtém métricas de desempenho de uma campanha.
        
        Args:
            campanha_id (str): ID da campanha
            data_inicio (datetime, optional): Data inicial para as métricas
            data_fim (datetime, optional): Data final para as métricas
            metricas (list, optional): Lista de métricas a serem obtidas
            
        Returns:
            dict: Métricas da campanha
        """
        try:
            # Configurar datas padrão se não fornecidas
            if not data_inicio:
                data_inicio = datetime.now() - timedelta(days=30)
            if not data_fim:
                data_fim = datetime.now()
            
            # Configurar métricas padrão se não fornecidas
            if not metricas:
                metricas = [
                    'impressions',
                    'clicks',
                    'ctr',
                    'spend',
                    'cpc',
                    'reach',
                    'frequency',
                ]
            
            # Obter as métricas
            campaign = Campaign(campanha_id)
            insights = campaign.get_insights(
                params={
                    'time_range': {
                        'since': data_inicio.strftime("%Y-%m-%d"),
                        'until': data_fim.strftime("%Y-%m-%d"),
                    },
                    'fields': metricas,
                }
            )
            
            # Processar e retornar os resultados
            if insights:
                return insights[0]
            else:
                return {}
        
        except FacebookRequestError as e:
            logger.error(f"Erro ao obter métricas da campanha: {e}")
            raise
    
    def buscar_interesses(self, termo_busca, limite=10):
        """
        Busca interesses para segmentação com base em um termo.
        
        Args:
            termo_busca (str): Termo para busca de interesses
            limite (int, optional): Número máximo de resultados. Default para 10
            
        Returns:
            list: Lista de interesses encontrados
        """
        try:
            # Buscar interesses
            interests = TargetingSearch.search(
                params={
                    'q': termo_busca,
                    'type': 'adinterest',
                    'limit': limite,
                }
            )
            
            # Processar e retornar os resultados
            return interests
        
        except FacebookRequestError as e:
            logger.error(f"Erro ao buscar interesses: {e}")
            raise
    
    def atualizar_status_campanha(self, campanha_id, status):
        """
        Atualiza o status de uma campanha.
        
        Args:
            campanha_id (str): ID da campanha
            status (str): Novo status (ACTIVE, PAUSED, ARCHIVED)
            
        Returns:
            bool: True se a atualização for bem-sucedida, False caso contrário
        """
        try:
            # Atualizar o status da campanha
            campaign = Campaign(campanha_id)
            campaign.update(params={'status': status})
            
            logger.info(f"Status da campanha {campanha_id} atualizado para {status}")
            return True
        
        except FacebookRequestError as e:
            logger.error(f"Erro ao atualizar status da campanha: {e}")
            return False 

File: lib/meta-ads-api.ts
import { initSDK, BizSdkInterface } from 'facebook-nodejs-business-sdk';

// Classes do SDK do Facebook
const { 
  FacebookAdsApi, 
  AdAccount, 
  Campaign, 
  AdSet, 
  Ad, 
  AdCreative, 
  TargetingSearch
} = require('facebook-nodejs-business-sdk') as BizSdkInterface;

// Carregar credenciais do ambiente
const META_APP_ID = process.env.META_APP_ID;
const META_APP_SECRET = process.env.META_APP_SECRET;
const META_ACCESS_TOKEN = process.env.META_ACCESS_TOKEN;
const META_ACCOUNT_ID = process.env.META_ACCOUNT_ID;

export interface CampaignParams {
  name: string;
  objective: string;
  dailyBudget?: number;
  lifetimeBudget?: number;
  startDate?: Date;
  endDate?: Date;
  status?: string;
  targetingSpec?: TargetingSpec;
}

export interface AdSetParams {
  name: string;
  campaignId: string;
  optimizationGoal: string;
  billingEvent?: string;
  bidAmount?: number;
  targetingSpec: TargetingSpec;
  dailyBudget?: number;
  lifetimeBudget?: number;
  startDate?: Date;
  endDate?: Date;
  status?: string;
}

export interface AdCreativeParams {
  name: string;
  adAccountId: string;
  title: string;
  body: string;
  imageUrl?: string;
  imageHash?: string;
  linkUrl: string;
  callToAction: string;
}

export interface AdParams {
  name: string;
  adsetId: string;
  creativeId: string;
  status?: string;
}

export interface TargetingSpec {
  age_min?: number;
  age_max?: number;
  genders?: number[];
  geo_locations?: {
    countries?: string[];
    cities?: any[];
    regions?: any[];
  };
  interests?: any[];
  behaviors?: any[];
  device_platforms?: string[];
  publisher_platforms?: string[];
  facebook_positions?: string[];
  instagram_positions?: string[];
}

export interface MetaAdsResponse {
  success: boolean;
  data?: any;
  error?: string;
  details?: any;
}

export interface CreateCampaignResponse extends MetaAdsResponse {
  data?: {
    campaignId?: string;
    adSetId?: string;
    adCreativeId?: string;
    adId?: string;
  };
}

/**
 * Classe para integração com a API do Meta Ads
 */
export class MetaAdsAPI {
  private api: typeof FacebookAdsApi;
  private adAccount: any;

  constructor() {
    if (!META_APP_ID || !META_APP_SECRET || !META_ACCESS_TOKEN || !META_ACCOUNT_ID) {
      throw new Error('Credenciais do Meta Ads não configuradas corretamente');
    }

    // Inicializar a API do Facebook
    this.api = FacebookAdsApi.init(META_ACCESS_TOKEN);
    this.adAccount = new AdAccount(`act_${META_ACCOUNT_ID}`);
    
    // Configurar modo de debug quando necessário
    if (process.env.NODE_ENV === 'development') {
      this.api.setDebug(true);
    }
  }

  /**
   * Cria uma campanha no Meta Ads
   */
  async createCampaign(params: CampaignParams): Promise<MetaAdsResponse> {
    try {
      // Validar parâmetros obrigatórios
      if (!params.name || !params.objective) {
        throw new Error('Nome e objetivo da campanha são obrigatórios');
      }

      // Configurar parâmetros da campanha
      const campaignParams: any = {
        name: params.name,
        objective: params.objective,
        status: params.status || 'PAUSED',
        special_ad_categories: [],
      };

      // Configurar orçamento (diário ou lifetime)
      if (params.dailyBudget) {
        campaignParams.daily_budget = Math.round(params.dailyBudget * 100); // Converter para centavos
      } else if (params.lifetimeBudget) {
        campaignParams.lifetime_budget = Math.round(params.lifetimeBudget * 100);
      }

      // Configurar datas
      if (params.startDate) {
        campaignParams.start_time = params.startDate.toISOString();
      }
      if (params.endDate) {
        campaignParams.end_time = params.endDate.toISOString();
      }

      // Criar campanha
      console.log('Criando campanha com parâmetros:', campaignParams);
      const campaign = await this.adAccount.createCampaign([], campaignParams);

      console.log('Campanha criada com sucesso:', campaign);
      return {
        success: true,
        data: {
          campaignId: campaign.id
        }
      };

    } catch (error: any) {
      console.error('Erro ao criar campanha:', error);
      return {
        success: false,
        error: 'Falha ao criar campanha',
        details: error.message || error
      };
    }
  }

  /**
   * Cria um conjunto de anúncios (AdSet) no Meta Ads
   */
  async createAdSet(params: AdSetParams): Promise<MetaAdsResponse> {
    try {
      // Validar parâmetros obrigatórios
      if (!params.name || !params.campaignId || !params.optimizationGoal || !params.targetingSpec) {
        throw new Error('Parâmetros insuficientes para criar conjunto de anúncios');
      }

      // Configurar parâmetros do conjunto de anúncios
      const adSetParams: any = {
        name: params.name,
        campaign_id: params.campaignId,
        optimization_goal: params.optimizationGoal,
        billing_event: params.billingEvent || 'IMPRESSIONS',
        status: params.status || 'PAUSED',
        targeting: params.targetingSpec,
      };

      // Configurar orçamento (diário ou lifetime)
      if (params.dailyBudget) {
        adSetParams.daily_budget = Math.round(params.dailyBudget * 100);
      } else if (params.lifetimeBudget) {
        adSetParams.lifetime_budget = Math.round(params.lifetimeBudget * 100);
      }

      // Configurar datas
      if (params.startDate) {
        adSetParams.start_time = params.startDate.toISOString();
      }
      if (params.endDate) {
        adSetParams.end_time = params.endDate.toISOString();
      }

      // Criar conjunto de anúncios
      console.log('Criando conjunto de anúncios com parâmetros:', adSetParams);
      const adSet = await this.adAccount.createAdSet([], adSetParams);

      console.log('Conjunto de anúncios criado com sucesso:', adSet);
      return {
        success: true,
        data: {
          adSetId: adSet.id
        }
      };

    } catch (error: any) {
      console.error('Erro ao criar conjunto de anúncios:', error);
      return {
        success: false,
        error: 'Falha ao criar conjunto de anúncios',
        details: error.message || error
      };
    }
  }

  /**
   * Cria um criativo de anúncio no Meta Ads
   */
  async createAdCreative(params: AdCreativeParams): Promise<MetaAdsResponse> {
    try {
      // Validar parâmetros obrigatórios
      if (!params.name || !params.title || !params.body || !params.linkUrl) {
        throw new Error('Parâmetros insuficientes para criar criativo');
      }

      // Obter o hash da imagem se for fornecida uma URL de imagem
      let imageHash = params.imageHash;
      if (!imageHash && params.imageUrl) {
        try {
          // Fazer upload da imagem
          const image = new AdImage(null, `act_${params.adAccountId}`);
          image.filename = params.imageUrl;
          await image.create();
          imageHash = image.hash;
        } catch (error) {
          console.error('Erro ao fazer upload de imagem:', error);
          // Continuar sem imagem
        }
      }

      // Configurar parâmetros do criativo
      const creativeParams: any = {
        name: params.name,
        object_story_spec: {
          page_id: process.env.META_PAGE_ID || '123456789', // ID da página do Facebook
          link_data: {
            message: params.body,
            link: params.linkUrl,
            name: params.title,
            call_to_action: {
              type: params.callToAction
            }
          }
        }
      };

      // Adicionar imagem se disponível
      if (imageHash) {
        creativeParams.object_story_spec.link_data.image_hash = imageHash;
      }

      // Criar criativo
      console.log('Criando criativo com parâmetros:', creativeParams);
      const creative = await this.adAccount.createAdCreative([], creativeParams);

      console.log('Criativo criado com sucesso:', creative);
      return {
        success: true,
        data: {
          adCreativeId: creative.id
        }
      };

    } catch (error: any) {
      console.error('Erro ao criar criativo:', error);
      return {
        success: false,
        error: 'Falha ao criar criativo',
        details: error.message || error
      };
    }
  }

  /**
   * Cria um anúncio no Meta Ads
   */
  async createAd(params: AdParams): Promise<MetaAdsResponse> {
    try {
      // Validar parâmetros obrigatórios
      if (!params.name || !params.adsetId || !params.creativeId) {
        throw new Error('Parâmetros insuficientes para criar anúncio');
      }

      // Configurar parâmetros do anúncio
      const adParams: any = {
        name: params.name,
        adset_id: params.adsetId,
        creative: {
          creative_id: params.creativeId
        },
        status: params.status || 'PAUSED'
      };

      // Criar anúncio
      console.log('Criando anúncio com parâmetros:', adParams);
      const ad = await this.adAccount.createAd([], adParams);

      console.log('Anúncio criado com sucesso:', ad);
      return {
        success: true,
        data: {
          adId: ad.id
        }
      };

    } catch (error: any) {
      console.error('Erro ao criar anúncio:', error);
      return {
        success: false,
        error: 'Falha ao criar anúncio',
        details: error.message || error
      };
    }
  }

  /**
   * Cria uma campanha completa com conjunto de anúncios e anúncio
   */
  async createFullCampaign(
    campaignParams: CampaignParams,
    adSetParams: Omit<AdSetParams, 'campaignId'>,
    creativeParams: AdCreativeParams,
    adParams: Omit<AdParams, 'adsetId' | 'creativeId'>
  ): Promise<CreateCampaignResponse> {
    try {
      // 1. Criar campanha
      const campaignResult = await this.createCampaign(campaignParams);
      if (!campaignResult.success) {
        return campaignResult;
      }
      const campaignId = campaignResult.data?.campaignId;

      // 2. Criar conjunto de anúncios
      const adSetResult = await this.createAdSet({
        ...adSetParams,
        campaignId
      });
      if (!adSetResult.success) {
        return {
          success: false,
          error: 'Falha ao criar conjunto de anúncios',
          details: adSetResult.details,
          data: {
            campaignId
          }
        };
      }
      const adSetId = adSetResult.data?.adSetId;

      // 3. Criar criativo
      const creativeResult = await this.createAdCreative({
        ...creativeParams,
        adAccountId: META_ACCOUNT_ID || ''
      });
      if (!creativeResult.success) {
        return {
          success: false,
          error: 'Falha ao criar criativo',
          details: creativeResult.details,
          data: {
            campaignId,
            adSetId
          }
        };
      }
      const adCreativeId = creativeResult.data?.adCreativeId;

      // 4. Criar anúncio
      const adResult = await this.createAd({
        ...adParams,
        adsetId: adSetId,
        creativeId: adCreativeId
      });
      if (!adResult.success) {
        return {
          success: false,
          error: 'Falha ao criar anúncio',
          details: adResult.details,
          data: {
            campaignId,
            adSetId,
            adCreativeId
          }
        };
      }
      const adId = adResult.data?.adId;

      // Retornar IDs de todos os componentes criados
      return {
        success: true,
        data: {
          campaignId,
          adSetId,
          adCreativeId,
          adId
        }
      };

    } catch (error: any) {
      console.error('Erro ao criar campanha completa:', error);
      return {
        success: false,
        error: 'Falha ao criar campanha completa',
        details: error.message || error
      };
    }
  }
} 